api_tests/test_extra_files.py

```diff
@@ -0,0 +1,74 @@
+"""
+Tests for uploading additional files and copying OS files into sandbox.
+
+Copying of python_path entries to the sandbox is performed by the webapp,
+and so AppArmor is not in effect. This means we have to take extra care here.
+"""
+
+import io
+import zipfile
+from textwrap import dedent
+
+from api_tests.utils import call_api_rejection, call_api_success
+
+
+def make_python_library_zip_bytes():
+    """
+    Make a simple course library ZIP file, returning the bytes.
+
+    Contains one module, `course_library`, with a function `triangular_number`.
+    """
+    memfile = io.BytesIO()
+    with zipfile.ZipFile(memfile, 'w') as z:
+        z.writestr(
+            'course_library.py',
+            dedent("""
+              def triangular_number(n):
+                  return sum(range(n + 1))
+            """),
+        )
+
+    memfile.seek(0)
+    return memfile.read()
+
+
+PYTHON_LIB_BYTES = make_python_library_zip_bytes()
+# Code to exercise the uploaded library.
+EXERCISE_LIBRARY = """
+from course_library import triangular_number
+out = triangular_number(6)
+"""
+# Expected globals output when the above code is run.
+EXPECTED_OUT = {'out': 21}
+
+
+def test_support_upload_file():
+    """
+    Can upload a python_lib.zip file and add it to the Python path.
+
+    This is the standard happy path. When a course has a `python_lib.zip` file,
+    edxapp automatically includes it this way.
+    """
+    globals_out = call_api_success(
+        EXERCISE_LIBRARY, {},
+        files={'python_lib.zip': PYTHON_LIB_BYTES}, python_path=['python_lib.zip'],
+    )
+    assert globals_out == EXPECTED_OUT
+
```

Comment:
You've got negative tests for both read and write, but a positive test only for read. Is there a reason for this? Or is there a missing test for a request to set files to `python_lib.zip`?

Reply:
The arbitrary read/arbitrary write tests are to test specific vulnerabilities (and the API's rejection of exploits) while this test is to check the happy path (upload one file under the "blessed" name, and use it). The happy path actually does both write and read.